torch/_inductor/codegen/cpp.py

```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
                 self._set_dependent_itervars(args[0])
             if any(arg.is_vec for arg in args if isinstance(arg, CppCSEVariable)):
                 self.is_vec = True
-        # NOTE [dtype of CppCSEVariable]
-        # Deciding dtype according to the current optimization context is not
-        # always accurate since the dtypes are initialized during dtype propagation
-        # at the beginning of the codegen. It is possible that some ops are invoked
-        # during the codegen of the current op and take different dtypes from the
-        # current op.
-        # TODO(jgong5): A more accurate way of deciding the dtype of the variables is to
-        # propagate the dtypes here inside `update_on_args`.
+        # NOTE [Deduce dtype of CppCSEVariable at runtime]
         if (
-            hasattr(V.interpreter, "current_node")
-            and get_current_node_opt_ctx() is not None
-        ):
+            output_dtype := _deduce_output_dtype_by_name(
+                name,
+                *args,
+                **kwargs,
+            )
+        ) is not None:
+            self.dtype = output_dtype
+        elif name == "masked":
+            assert (
+                hasattr(V.interpreter, "current_node")
+                and V.interpreter.current_node.target.startswith("masked_subblock")
+                and get_current_node_opt_ctx() is not None
+            )
             self.dtype = get_current_node_opt_ctx().dtype
-
-        if name in BIN_CMP_OPS:
-            self.dtype = torch.bool
+        else:
+            # Induce output dtype by inputs' dtype
+            assert all(
+                arg.dtype is not None for arg in args if isinstance(arg, CppCSEVariable)
+            )
+            self.dtype = functools.reduce(
+                torch.promote_types,  # type: ignore[arg-type]
+                [arg.dtype for arg in args if isinstance(arg, CppCSEVariable)],
+            )
```

Comment:
Can we move this logic into `_deduce_output_dtype_by_name`?

Reply:
Thanks for the comment. As synced offline, move it into a util function inside `cpp_utils.py`